apps/remote_control/lib/lexical/remote_control/code_intelligence/rename/module.ex

```diff
@@ -0,0 +1,164 @@
+defmodule Lexical.RemoteControl.CodeIntelligence.Rename.Module do
+  alias Lexical.Ast
+  alias Lexical.Ast.Analysis
+  alias Lexical.Document
+  alias Lexical.Document.Edit
+  alias Lexical.Document.Line
+  alias Lexical.Document.Position
+  alias Lexical.Document.Range
+  alias Lexical.RemoteControl.CodeIntelligence.Entity
+  alias Lexical.RemoteControl.Search.Store
+  require Logger
+
+  import Line
+
+  @spec rename(Analysis.t(), Position.t(), String.t()) ::
+          {:ok, %{Lexical.uri() => [Edit.t()]}} | {:error, term()}
+  def rename(%Analysis{} = analysis, %Position{} = position, new_name) do
+    with {:ok, entity, range} <- resolve_module(analysis, position) do
+      edits =
+        analysis.document
+        |> search_related_candidates(position, entity, range)
+        |> to_edits_by_uri(new_name)
+
+      {:ok, edits}
+    end
+  end
+
+  @spec prepare(Analysis.t(), Position.t()) :: {:ok, String.t(), Range.t()} | {:error, term()}
+  def prepare(%Analysis{} = analysis, %Position{} = position) do
+    case resolve_module(analysis, position) do
+      {:ok, _, range} ->
+        {:ok, local_module_name(range), range}
+
+      {:error, _} ->
+        {:error, :unsupported_entity}
+    end
+  end
+
+  defp resolve_module(analysis, position) do
+    case Entity.resolve(analysis, position) do
+      {:ok, {module_or_struct, module}, range} when module_or_struct in [:struct, :module] ->
+        {:ok, module, range}
+
+      {:ok, other, _} ->
+        {:error, {:unsupported_entity, other}}
+
+      {:error, reason} ->
+        {:error, reason}
+    end
+  end
+
+  defp search_related_candidates(document, position, entity, range) do
+    local_module_name = local_module_name(range)
+    entities = exacts(entity, local_module_name)
+
+    # Users won't always want to rename descendants of a module.
+    # For instance, when there are no more submodules after the cursor.
+    # like: `defmodule TopLevel.Mo|dule do`
+    # in most cases, users only want to rename the module itself.
```

Comment:
I think you make a fair point, and I need to give more thought to the ergonomics aspect.

Additionally, a part that this PR is missing is that it currently only allows for renaming the module directly **under the cursor** (between two points, or between a point and a space, etc). If we could support renaming the entire module completely, it might be more effective, though this could also be more challenging.

Reply:
has been fixed, the new method is much simpler.